File: HeapDataStructure.java
import java.io.*;
import java.util.*;

class Node {
    int id;
    String nama;
    
    Node(int id, String nama) {
        this.id = id;
        this.nama = nama;
    }
}

class MinHeap {
    ArrayList<Node> heap;
    
    MinHeap() {
        heap = new ArrayList<>();
    }
    
    // Tambah data ke min-heap
    void insert(int id, String nama) {
        heap.add(new Node(id, nama));
        heapifyUp(heap.size() - 1);
    }
    
    void heapifyUp(int idx) {
        int parent = (idx - 1) / 2;
        if (idx > 0 && heap.get(parent).id > heap.get(idx).id) {
            Node temp = heap.get(parent);
            heap.set(parent, heap.get(idx));
            heap.set(idx, temp);
            heapifyUp(parent);
        }
    }
    
    void heapifyDown(int idx) {
        int smallest = idx;
        int left = 2 * idx + 1;
        int right = 2 * idx + 2;
        
        if (left < heap.size() && heap.get(left).id < heap.get(smallest).id) {
            smallest = left;
        }
        if (right < heap.size() && heap.get(right).id < heap.get(smallest).id) {
            smallest = right;
        }
        
        if (smallest != idx) {
            Node temp = heap.get(idx);
            heap.set(idx, heap.get(smallest));
            heap.set(smallest, temp);
            heapifyDown(smallest);
        }
    }
    
    // Hapus data dari min-heap berdasarkan ID
    boolean deleteById(int id) {
        for (int i = 0; i < heap.size(); i++) {
            if (heap.get(i).id == id) {
                heap.set(i, heap.get(heap.size() - 1));
                heap.remove(heap.size() - 1);
                heapifyDown(i);
                heapifyUp(i);
                return true;
            }
        }
        return false;
    }
    
    // Mendapatkan data urut ascending (tanpa merusak heap asli)
    ArrayList<Node> getSorted() {
        ArrayList<Node> temp = new ArrayList<>();
        for (Node n : heap) {
            temp.add(new Node(n.id, n.nama));
        }
        
        ArrayList<Node> result = new ArrayList<>();
        
        while (!temp.isEmpty()) {
            // Extract min
            result.add(temp.get(0));
            temp.set(0, temp.get(temp.size() - 1));
            temp.remove(temp.size() - 1);
            
            int idx = 0;
            while (true) {
                int smallest = idx;
                int left = 2 * idx + 1;
                int right = 2 * idx + 2;
                
                if (left < temp.size() && temp.get(left).id < temp.get(smallest).id) {
                    smallest = left;
                }
                if (right < temp.size() && temp.get(right).id < temp.get(smallest).id) {
                    smallest = right;
                }
                
                if (smallest != idx) {
                    Node swap = temp.get(idx);
                    temp.set(idx, temp.get(smallest));
                    temp.set(smallest, swap);
                    idx = smallest;
                } else {
                    break;
                }
            }
        }
        return result;
    }
}

class MaxHeap {
    ArrayList<Node> heap;
    
    MaxHeap() {
        heap = new ArrayList<>();
    }
    
    // Tambah data ke max-heap
    void insert(int id, String nama) {
        heap.add(new Node(id, nama));
        heapifyUp(heap.size() - 1);
    }
    
    void heapifyUp(int idx) {
        int parent = (idx - 1) / 2;
        if (idx > 0 && heap.get(parent).id < heap.get(idx).id) {
            Node temp = heap.get(parent);
            heap.set(parent, heap.get(idx));
            heap.set(idx, temp);
            heapifyUp(parent);
        }
    }
    
    void heapifyDown(int idx) {
        int largest = idx;
        int left = 2 * idx + 1;
        int right = 2 * idx + 2;
        
        if (left < heap.size() && heap.get(left).id > heap.get(largest).id) {
            largest = left;
        }
        if (right < heap.size() && heap.get(right).id > heap.get(largest).id) {
            largest = right;
        }
        
        if (largest != idx) {
            Node temp = heap.get(idx);
            heap.set(idx, heap.get(largest));
            heap.set(largest, temp);
            heapifyDown(largest);
        }
    }
    
    // Hapus data dari max-heap berdasarkan ID
    boolean deleteById(int id) {
        for (int i = 0; i < heap.size(); i++) {
            if (heap.get(i).id == id) {
                heap.set(i, heap.get(heap.size() - 1));
                heap.remove(heap.size() - 1);
                heapifyDown(i);
                heapifyUp(i);
                return true;
            }
        }
        return false;
    }
    
    // Mendapatkan data urut descending (tanpa merusak heap asli)
    ArrayList<Node> getSorted() {
        ArrayList<Node> temp = new ArrayList<>();
        for (Node n : heap) {
            temp.add(new Node(n.id, n.nama));
        }
        
        ArrayList<Node> result = new ArrayList<>();
        
        while (!temp.isEmpty()) {
            // Extract max
            result.add(temp.get(0));
            temp.set(0, temp.get(temp.size() - 1));
            temp.remove(temp.size() - 1);
            
            int idx = 0;
            while (true) {
                int largest = idx;
                int left = 2 * idx + 1;
                int right = 2 * idx + 2;
                
                if (left < temp.size() && temp.get(left).id > temp.get(largest).id) {
                    largest = left;
                }
                if (right < temp.size() && temp.get(right).id > temp.get(largest).id) {
                    largest = right;
                }
                
                if (largest != idx) {
                    Node swap = temp.get(idx);
                    temp.set(idx, temp.get(largest));
                    temp.set(largest, swap);
                    idx = largest;
                } else {
                    break;
                }
            }
        }
        return result;
    }
}

public class HeapDataStructure {
    MinHeap minHeap;
    MaxHeap maxHeap;
    
    HeapDataStructure() {
        minHeap = new MinHeap();
        maxHeap = new MaxHeap();
    }
    
    // (1) Tambah data ke min-heap dan max-heap sekaligus
    void tambahData(int id, String nama) {
        minHeap.insert(id, nama);
        maxHeap.insert(id, nama);
        System.out.println("✓ Data berhasil ditambahkan: (" + id + ", " + nama + ")");
    }
    
    // (2) Tampilkan data ascending menggunakan Min-Heap
    void tampilAscending() {
        if (minHeap.heap.isEmpty()) {
            System.out.println("\n⚠ Min-Heap kosong!");
            return;
        }
        System.out.println("\n=== DATA URUT ASCENDING (MIN-HEAP) ===");
        System.out.println("ID\t\tNama");
        System.out.println("--------------------------------");
        ArrayList<Node> sortedData = minHeap.getSorted();
        for (Node node : sortedData) {
            System.out.println(node.id + "\t\t" + node.nama);
        }
    }
    
    // (3) Tampilkan data descending menggunakan Max-Heap
    void tampilDescending() {
        if (maxHeap.heap.isEmpty()) {
            System.out.println("\n⚠ Max-Heap kosong!");
            return;
        }
        System.out.println("\n=== DATA URUT DESCENDING (MAX-HEAP) ===");
        System.out.println("ID\t\tNama");
        System.out.println("--------------------------------");
        ArrayList<Node> sortedData = maxHeap.getSorted();
        for (Node node : sortedData) {
            System.out.println(node.id + "\t\t" + node.nama);
        }
    }
    
    // (4) Hapus data dari min-heap
    void hapusDariMinHeap(int id) {
        if (minHeap.deleteById(id)) {
            System.out.println("✓ Data ID " + id + " berhasil dihapus dari Min-Heap");
        } else {
            System.out.println("✗ Data ID " + id + " tidak ditemukan di Min-Heap");
        }
    }
    
    // (5) Hapus data dari max-heap
    void hapusDariMaxHeap(int id) {
        if (maxHeap.deleteById(id)) {
            System.out.println("✓ Data ID " + id + " berhasil dihapus dari Max-Heap");
        } else {
            System.out.println("✗ Data ID " + id + " tidak ditemukan di Max-Heap");
        }
    }
    
    // Load data awal dari CSV (path langsung di kodingan)
    void loadFromCSV() {
        String filename = "/Users/clumsy/coding/struktur-data-py-java/HeapDataStructure/data.csv";
        int count = 0;
        
        try (BufferedReader br = new BufferedReader(new FileReader(filename))) {
            String line;
            boolean firstLine = true;
            
            while ((line = br.readLine()) != null) {
                if (firstLine) {
                    firstLine = false;
                    continue;
                }
                if (line.trim().isEmpty()) continue;
                
                String[] data = line.split(",");
                if (data.length >= 2) {
                    try {
                        int id = Integer.parseInt(data[0].trim());
                        String nama = data[1].trim();
                        tambahData(id, nama);
                        count++;
                    } catch (NumberFormatException e) {
                        System.out.println("Baris tidak valid: " + line);
                    }
                }
            }
            System.out.println("\n✅ Data berhasil dimuat dari " + filename);
            System.out.println("Total data: " + count);
        } catch (FileNotFoundException e) {
            System.out.println("\n❌ File tidak ditemukan: " + filename);
        } catch (Exception e) {
            System.out.println("Error: " + e.getMessage());
        }
    }
    
    public static void main(String[] args) {
        HeapDataStructure heap = new HeapDataStructure();
        Scanner scanner = new Scanner(System.in);
        
        // Starting data awal - load dari CSV
        System.out.println("Memuat data dari file CSV...");
        heap.loadFromCSV();
        
        while (true) {
            System.out.println("\n╔════════════════════════════════════════╗");
            System.out.println("║          HEAP DATA STRUCTURE           ║");
            System.out.println("╠════════════════════════════════════════╣");
            System.out.println("║ 1. Tambah data ke Min-Heap & Max-Heap  ║");
            System.out.println("║ 2. Tampil ascending (Min-Heap)         ║");
            System.out.println("║ 3. Tampil descending (Max-Heap)        ║");
            System.out.println("║ 4. Hapus data dari Min-Heap            ║");
            System.out.println("║ 5. Hapus data dari Max-Heap            ║");
            System.out.println("║ 0. Keluar                              ║");
            System.out.println("╚════════════════════════════════════════╝");
            System.out.print("Pilihan: ");
            
            String pilihan = scanner.nextLine().trim();
            
            if (pilihan.equals("1")) {
                try {
                    System.out.print("Masukkan ID: ");
                    int id = Integer.parseInt(scanner.nextLine());
                    System.out.print("Masukkan Nama: ");
                    String nama = scanner.nextLine();
                    heap.tambahData(id, nama);
                } catch (NumberFormatException e) {
                    System.out.println("ID harus berupa angka!");
                }
            } else if (pilihan.equals("2")) {
                heap.tampilAscending();
            } else if (pilihan.equals("3")) {
                heap.tampilDescending();
            } else if (pilihan.equals("4")) {
                try {
                    System.out.print("Masukkan ID yang akan dihapus dari Min-Heap: ");
                    int id = Integer.parseInt(scanner.nextLine());
                    heap.hapusDariMinHeap(id);
                } catch (NumberFormatException e) {
                    System.out.println("ID harus berupa angka!");
                }
            } else if (pilihan.equals("5")) {
                try {
                    System.out.print("Masukkan ID yang akan dihapus dari Max-Heap: ");
                    int id = Integer.parseInt(scanner.nextLine());
                    heap.hapusDariMaxHeap(id);
                } catch (NumberFormatException e) {
                    System.out.println("ID harus berupa angka!");
                }
            } else if (pilihan.equals("0")) {
                System.out.println("Terima kasih!");
                break;
            } else {
                System.out.println("Pilihan tidak valid!");
            }
        }
        
        scanner.close();
    }
}

File: HeapDataStructure.py
import csv

class Node:
    def __init__(self, id, nama):
        self.id = id
        self.nama = nama

class MinHeap:
    def __init__(self):
        self.heap = []
    
    def insert(self, id, nama):
        """Tambah data ke min-heap"""
        self.heap.append(Node(id, nama))
        self._heapify_up(len(self.heap) - 1)
    
    def _heapify_up(self, idx):
        parent = (idx - 1) // 2
        if idx > 0 and self.heap[parent].id > self.heap[idx].id:
            self.heap[parent], self.heap[idx] = self.heap[idx], self.heap[parent]
            self._heapify_up(parent)
    
    def _heapify_down(self, idx):
        smallest = idx
        left = 2 * idx + 1
        right = 2 * idx + 2
        
        if left < len(self.heap) and self.heap[left].id < self.heap[smallest].id:
            smallest = left
        if right < len(self.heap) and self.heap[right].id < self.heap[smallest].id:
            smallest = right
        
        if smallest != idx:
            self.heap[idx], self.heap[smallest] = self.heap[smallest], self.heap[idx]
            self._heapify_down(smallest)
    
    def delete_by_id(self, id):
        """Hapus data dari min-heap berdasarkan ID (PERBAIKAN)"""
        # Cari index data yang mau dihapus
        index_to_delete = -1
        for i in range(len(self.heap)):
            if self.heap[i].id == id:
                index_to_delete = i
                break
        
        if index_to_delete == -1:
            return False
        
        # Swap dengan data terakhir
        last_index = len(self.heap) - 1
        self.heap[index_to_delete], self.heap[last_index] = self.heap[last_index], self.heap[index_to_delete]
        
        # Hapus data terakhir
        self.heap.pop()
        
        # Perbaiki heap jika masih ada data
        if index_to_delete < len(self.heap):
            self._heapify_down(index_to_delete)
            self._heapify_up(index_to_delete)
        
        return True
    
    def get_sorted(self):
        """Mendapatkan data urut ascending (tanpa merusak heap asli)"""
        if not self.heap:
            return []
        
        temp = self.heap.copy()
        result = []
        while temp:
            # Extract min
            result.append(temp[0])
            temp[0] = temp[-1]
            temp.pop()
            idx = 0
            while True:
                smallest = idx
                left = 2 * idx + 1
                right = 2 * idx + 2
                if left < len(temp) and temp[left].id < temp[smallest].id:
                    smallest = left
                if right < len(temp) and temp[right].id < temp[smallest].id:
                    smallest = right
                if smallest != idx:
                    temp[idx], temp[smallest] = temp[smallest], temp[idx]
                    idx = smallest
                else:
                    break
        return result

class MaxHeap:
    def __init__(self):
        self.heap = []
    
    def insert(self, id, nama):
        """Tambah data ke max-heap"""
        self.heap.append(Node(id, nama))
        self._heapify_up(len(self.heap) - 1)
    
    def _heapify_up(self, idx):
        parent = (idx - 1) // 2
        if idx > 0 and self.heap[parent].id < self.heap[idx].id:
            self.heap[parent], self.heap[idx] = self.heap[idx], self.heap[parent]
            self._heapify_up(parent)
    
    def _heapify_down(self, idx):
        largest = idx
        left = 2 * idx + 1
        right = 2 * idx + 2
        
        if left < len(self.heap) and self.heap[left].id > self.heap[largest].id:
            largest = left
        if right < len(self.heap) and self.heap[right].id > self.heap[largest].id:
            largest = right
        
        if largest != idx:
            self.heap[idx], self.heap[largest] = self.heap[largest], self.heap[idx]
            self._heapify_down(largest)
    
    def delete_by_id(self, id):
        """Hapus data dari max-heap berdasarkan ID (PERBAIKAN)"""
        # Cari index data yang mau dihapus
        index_to_delete = -1
        for i in range(len(self.heap)):
            if self.heap[i].id == id:
                index_to_delete = i
                break
        
        if index_to_delete == -1:
            return False
        
        # Swap dengan data terakhir
        last_index = len(self.heap) - 1
        self.heap[index_to_delete], self.heap[last_index] = self.heap[last_index], self.heap[index_to_delete]
        
        # Hapus data terakhir
        self.heap.pop()
        
        # Perbaiki heap jika masih ada data
        if index_to_delete < len(self.heap):
            self._heapify_down(index_to_delete)
            self._heapify_up(index_to_delete)
        
        return True
    
    def get_sorted(self):
        """Mendapatkan data urut descending (tanpa merusak heap asli)"""
        if not self.heap:
            return []
        
        temp = self.heap.copy()
        result = []
        while temp:
            # Extract max
            result.append(temp[0])
            temp[0] = temp[-1]
            temp.pop()
            idx = 0
            while True:
                largest = idx
                left = 2 * idx + 1
                right = 2 * idx + 2
                if left < len(temp) and temp[left].id > temp[largest].id:
                    largest = left
                if right < len(temp) and temp[right].id > temp[largest].id:
                    largest = right
                if largest != idx:
                    temp[idx], temp[largest] = temp[largest], temp[idx]
                    idx = largest
                else:
                    break
        return result

class HeapDataStructure:
    def __init__(self):
        self.minHeap = MinHeap()
        self.maxHeap = MaxHeap()
    
    def tambahData(self, id, nama):
        """Tambah data ke min-heap dan max-heap sekaligus"""
        self.minHeap.insert(id, nama)
        self.maxHeap.insert(id, nama)
        print(f"✓ Data berhasil ditambahkan: ({id}, {nama})")
    
    def tampilAscending(self):
        """Tampilkan data ascending menggunakan Min-Heap"""
        if not self.minHeap.heap:
            print("\n⚠ Min-Heap kosong!")
            return
        print("\n=== DATA URUT ASCENDING (MIN-HEAP) ===")
        print("ID\t\tNama")
        print("--------------------------------")
        sorted_data = self.minHeap.get_sorted()
        for node in sorted_data:
            print(f"{node.id}\t\t{node.nama}")
    
    def tampilDescending(self):
        """Tampilkan data descending menggunakan Max-Heap"""
        if not self.maxHeap.heap:
            print("\n⚠ Max-Heap kosong!")
            return
        print("\n=== DATA URUT DESCENDING (MAX-HEAP) ===")
        print("ID\t\tNama")
        print("--------------------------------")
        sorted_data = self.maxHeap.get_sorted()
        for node in sorted_data:
            print(f"{node.id}\t\t{node.nama}")
    
    def hapusDariMinHeap(self, id):
        """Hapus data dari min-heap"""
        if self.minHeap.delete_by_id(id):
            # Hapus juga dari max-heap biar konsisten
            self.maxHeap.delete_by_id(id)
            print(f"✓ Data ID {id} berhasil dihapus dari Min-Heap & Max-Heap")
        else:
            print(f"✗ Data ID {id} tidak ditemukan")
    
    def hapusDariMaxHeap(self, id):
        """Hapus data dari max-heap"""
        if self.maxHeap.delete_by_id(id):
            # Hapus juga dari min-heap biar konsisten
            self.minHeap.delete_by_id(id)
            print(f"✓ Data ID {id} berhasil dihapus dari Max-Heap & Min-Heap")
        else:
            print(f"✗ Data ID {id} tidak ditemukan")
    
    def loadFromCSV(self):
        """Load data awal dari CSV (path langsung di kodingan)"""
        filename = "/Users/clumsy/coding/struktur-data-py-java/HeapDataStructure/data.csv"
        count = 0
        try:
            with open(filename, 'r') as file:
                csv_reader = csv.reader(file)
                firstLine = True
                for row in csv_reader:
                    if firstLine:
                        firstLine = False
                        continue
                    if len(row) >= 2 and row[0].strip():
                        id = int(row[0].strip())
                        nama = row[1].strip()
                        self.tambahData(id, nama)
                        count += 1
            print(f"\n✅ Data berhasil dimuat dari {filename}")
            print(f"Total data: {count}")
        except FileNotFoundError:
            print(f"\n❌ File tidak ditemukan: {filename}")
        except Exception as e:
            print(f"Error: {e}")

def main():
    heap = HeapDataStructure()
    
    print("Memuat data dari file CSV...")
    heap.loadFromCSV()
    
    while True:
        print("\n╔════════════════════════════════════════╗")
        print("║          HEAP DATA STRUCTURE           ║")
        print("╠════════════════════════════════════════╣")
        print("║ 1. Tambah data ke Min-Heap & Max-Heap  ║")
        print("║ 2. Tampil ascending (Min-Heap)         ║")
        print("║ 3. Tampil descending (Max-Heap)        ║")
        print("║ 4. Hapus data dari Min-Heap            ║")
        print("║ 5. Hapus data dari Max-Heap            ║")
        print("║ 0. Keluar                              ║")
        print("╚════════════════════════════════════════╝")
        
        pilihan = input("Pilihan: ")
        
        if pilihan == '1':
            try:
                id = int(input("Masukkan ID: "))
                nama = input("Masukkan Nama: ")
                heap.tambahData(id, nama)
            except ValueError:
                print("ID harus berupa angka!")
        elif pilihan == '2':
            heap.tampilAscending()
        elif pilihan == '3':
            heap.tampilDescending()
        elif pilihan == '4':
            try:
                id = int(input("Masukkan ID yang akan dihapus dari Min-Heap: "))
                heap.hapusDariMinHeap(id)
            except ValueError:
                print("ID harus berupa angka!")
        elif pilihan == '5':
            try:
                id = int(input("Masukkan ID yang akan dihapus dari Max-Heap: "))
                heap.hapusDariMaxHeap(id)
            except ValueError:
                print("ID harus berupa angka!")
        elif pilihan == '0':
            print("Terima kasih!")
            break
        else:
            print("Pilihan tidak valid!")

if __name__ == "__main__":
    main()
